lib/pipelines/code-pipeline.ts

```diff
@@ -2,57 +2,166 @@ import * as cdk from '@aws-cdk/core';
 import * as pipelines from '@aws-cdk/pipelines';
 import * as codepipeline from '@aws-cdk/aws-codepipeline';
 import * as actions from '@aws-cdk/aws-codepipeline-actions';
+import { Construct, StackProps } from '@aws-cdk/core';
+import { ApplicationRepository, AsyncStackBuilder, StackBuilder } from '../spi';
+import { AddStageOptions } from '@aws-cdk/pipelines';
+import { withUsageTracking } from '../utils/usage-utils'; 
+
+
+/**
+ * credentialsType is excluded and the only supported credentialsSecret is a plaintext GitHub OAuth token.
+ */
+export type GitHubSourceRepository = Omit<ApplicationRepository, "credentialsType">;
 
 /**
  * Props for the Pipeline.
  */
 export type PipelineProps = {
+    
     /**
      * The name for the pipeline.
      */
-    name: string
+    name: string;
 
     /**
      * The owner of the repository for the pipeline (GitHub handle).
      */
-    owner: string
+    owner: string;
+
+    /**
+     * Repository for the pipeline
+     */
+    repository: GitHubSourceRepository;
 
     /**
-     * The name of the repository for the pipeline.
+     * Pipeline stages and options.
      */
-    repo: string
+    stages: StackStage[];
+}
 
+/**
+ * Stack stage is a builder construct to allow adding stages to a pipeline. Each stage is expected to produce a stack. 
+ */
+export interface StackStage {
     /**
-     * The branch for the pipeline.
+     * id of the stage
      */
-    branch: string
+    id: string;
 
     /**
-     * Secret key for GitHub OAuth credentials (stored in SecretsManager).
+     * Builder that can produce a stack which will be deployed as part of the stage
      */
-    secretKey?: string
+    stackBuilder: StackBuilder;
 
     /**
-     * The CDK scope for the pipeline.
+     * Optional stage properties, such as {manualApprovals: true} which can control stage transitions.
      */
-    scope: cdk.Construct
+    stageProps?: AddStageOptions
 }
 
 /**
- * CodePipeline deploys a new CodePipeline resource that is integrated with a GitHub repository.
+ * Builder for CodePipeline.
  */
-export class CodePipeline {
+export class CodePipelineBuilder implements StackBuilder {
+
+    private props: Partial<PipelineProps>;
 
-    public static build = (props: PipelineProps) => {
+
+    constructor() {
+        this.props = { stages: []};
+    }
+
+    public name(name: string): CodePipelineBuilder {
```

Comment:
are these meant to be setters? If so, consider renaming to `setName`. Also what does it return this?

Reply:
They are not traditional setters. That the convention for the builder pattern. The difference is fluid API as each method returns `this`.  Alternative is adding `with` to each property but that seems overly verbose for little gain.

Reply:
gotcha. My vote would be to use setter function names. Makes it explicit that you are "setting" something as opposed to reading. Up to you.

Reply:
Builders are used to reduce verbosity as one of the objectives and extra prefixes are not quite helping. I am leaning towards keeping it as is, even though i understand your perspective.